### PR TITLE
Add a page explaining script purposes

### PR DESCRIPTION
**Pull request by zliu41**
It explains what each variant of `ScriptPurpose` does.

### CODE REVIEW

**Review by zeme-wana — APPROVED** [review]

**Review by ana-pantilie — APPROVED** [review]
Thank you for writing this, it's very informative!

File: doc/docusaurus/docs/working-with-scripts/script-purposes.md [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+## Spending
+
+A spending script validates the spending of a UTXO.
+A UTXO can have either a public key address or a script address.
+To spend a UTXO with a script address, the script whose hash matches the script address must be included in the transaction (either directly, or as a reference script), and is executed to validate it.
+
+The script can make the decision based on the datum attached in the UTXO being spent, if any (datum is mandatory for Plutus V1 and V2, i.e., a UTXO with a Plutus V1 or V2 script address but without a datum cannot be spent; datum is optional for Plutus V3), the redeemer included in the transaction, as well as other fields of the transaction.
+
+The `Spending` constructor of `ScriptPurpose` contains a `TxOutRef` field, which is the UTXO the transaction is attempting to spend.
```
Comment (rober-m) [inline doc/docusaurus/docs/working-with-scripts/script-purposes.md:23]: I think this is a bit clearer (just a suggestion):

"The `Spending` constructor of `ScriptPurpose` includes a `TxOutRef` field that references the UTXO the script is validating, which is also one of the UTXOs the transaction attempts to spend."

File: doc/docusaurus/docs/working-with-scripts/script-purposes.md [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+## Minting
+
+Minting scripts, sometimes also referred to as minting policies, are used to approve or reject minting of new assets.
+
+An asset on Cardano contains a `CurrencySymbol` and a `TokenName`.
```
Comment (rober-m) [inline doc/docusaurus/docs/working-with-scripts/script-purposes.md:29]: Suggestion to improve clarity (especially for newcomers):

"In Cardano, we can uniquely identify an asset by its `CurrencySymbol` and `TokenName.`"

File: doc/docusaurus/docs/working-with-scripts/script-purposes.md [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+Staking rewards are deposited into a reward account[^1] corresponding to the staking credential.
+
+A staking credential can contain either a public key hash or a script hash.
+To withdraw rewards from a reward account corresponding to a staking credential that contains a script hash, the script with that particular hash must be included in the transaction, and is executed to validate the withdrawal.
+
+For example, such a script may check that any withdrawal must not take more than 10 Ada at a time, and that any withdrawal must be deposited into one of the allowed addresses.
+
+## Certifying
+
+A certifying script validates one of the following actions a transaction attempts to make: (1) registering a staking credential, and in doing so, creating a reward account associated with the staking credential; (2) de-registering a staking credential, and in doing so, terminating the reward account; (3) delegating a staking credential to a particular delegatee.
```
Comment (rober-m) [inline doc/docusaurus/docs/working-with-scripts/script-purposes.md:50]: There are more actions related to certificates (see [haddock](https://plutus.cardano.intersectmbo.org/haddock/latest/plutus-ledger-api/PlutusLedgerApi-V3.html#t:TxCert)). But maybe it'll be too much to go over each one. We could do something like:

"A certifying script can validate many certificate-related transactions. For example:"

That suggests there are more certificates without needing a significant rewrite of this section.

File: doc/docusaurus/docs/working-with-scripts/script-purposes.md [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+As previously stated, a UTXO's address can be either a public key address or a script address.
+Both kinds of addresses can optionally have a staking credential.
+A UTXO may contain Ada, and the Ada it contains can be delegated to an SPO to earn staking rewards.
+Staking rewards are deposited into a reward account[^1] corresponding to the staking credential.
+
+A staking credential can contain either a public key hash or a script hash.
+To withdraw rewards from a reward account corresponding to a staking credential that contains a script hash, the script with that particular hash must be included in the transaction, and is executed to validate the withdrawal.
+
+For example, such a script may check that any withdrawal must not take more than 10 Ada at a time, and that any withdrawal must be deposited into one of the allowed addresses.
```
[the remark this answers is not in the document]
Comment (CarlosLopezDeLara) [reply doc/docusaurus/docs/working-with-scripts/script-purposes.md:46]: @rober-m is correct: Partial withdrawals are not allowed. 

If the second part of the example is possible: _any withdrawal must be deposited into one of the allowed addresses_, this alone would be a great example.

File: doc/docusaurus/docs/working-with-scripts/script-purposes.md [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+In all three cases, if the staking credential in question contains a script hash (as opposed to a public key hash), the script with that hash must be included in the transaction, and is executed to validate the action.
+
+Such a script may, for instance, check that certain signatures be provided for registration, de-registration and delegation, or that the delegatee must be among the allowed delegatees.
+
+## Voting
+
+A voting script validates votes cast by a DRep or a constitutional committee member in a transaction.
+A DRep or a constitution committee member can be associated with a script hash.
+If a transaction contains one or more votes from a DRep or a constitution committee member associated with a script hash, the script with that hash must be included in the transaction, and is executed to approve or reject the vote.
+
```
Comment (CarlosLopezDeLara) [inline doc/docusaurus/docs/working-with-scripts/script-purposes.md:61]: I think it is better to use full name on the first occurrence, and use the abbreviation, acronym or initialism. i.e. 

"A voting script validates votes cast by a Delegated Representative (DRep) or a Constitutional committee member (CCM) in a transaction.

A DRep or a CCM can be associated with a script hash..."

File: doc/docusaurus/docs/working-with-scripts/script-purposes.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (zliu41) [reply doc/docusaurus/docs/working-with-scripts/script-purposes.md:46]: Good catch, and good to know!

**Pull request merged**